WordPress/gutenberg
# Introduce `withSyncEvent` action wrapper utility and proxy `event` object whenever it is not used

This implements the 1st step from #64944. Please see the issue for additional context on the rationale and overarching approach for this.

## Relevant technical choices

* In order for directive implementations to access resolved data from the entries, a few new functions are needed. This is because the `evaluate` function resolves the entry into the relevant data, but then also immediately evaluates it.
    * For the case of an event listener callback, this means the `evaluate` function will not only "look up" the callback, but also call it.
    * Since we need to conditionally alter the `event` object based on whether the callback is wrapped in `withSyncEvent` prior to calling the callback, the `evaluate` function is not sufficient.
    * In a future PR, we would then also need to call `splitTask()` before running the callback function (see 2nd step from #64944).
* As a solution, this PR introduces two new functions `resolveEntry` and `evaluateResolved`. If they are called right after each other, it's effectively the same as `evaluate`. In fact, this PR updates `evaluate` to rely on these two lower-level functions, to avoid duplicate code.
* Most directives can still use `evaluate` as is, which is more ergonomic. However the `on` (but not `on-async`) directives require the more granular access of the callback first, to only execute it later.

## Note to reviewers

Because several files are touched, but only _some_ specific changes are non-trivial, I would advise to review the commits individually at first. That should help convey the bigger picture than looking at the whole PR diff at once.

## Testing instructions

1. Add an image block to a post or page and enable the lightbox for it.
2. In the frontend, after opening the lightbox, navigate using the `Tab` key and ensure no warnings are present in the browser console.
    * This is because the `handleKeydown` action of the `core/image` block's store passes its callback through `withSyncEvent`.
3. You can now alter the code to see what would happen if the `handleKeydown` action was doing it wrong: Edit `packages/block-library/src/image/view.js` and remove the `withSyncEvent` usage from the `handleKeydown` callback, so that it is no longer wrapped. Don't forget to rerun the JS build.
4. Repeat the frontend interaction from step 2, and you should see a warning in the browser console that `withSyncEvent` needs to be used.
5. Add a search block to a post or page and set its button position to "button only".
6. In the frontend, click the Search button and ensure no warnings are present in the browser console.
    * This is because the `openSearchInput` action of the `core/search` block's store passes its callback through `withSyncEvent`.
7. Similar to step 3, edit `packages/block-library/src/search/view.js` and remove the `withSyncEvent` usage from the `openSearchInput` callback, so that it is no longer wrapped. Don't forget to rerun the JS build.
8. Repeat the frontend interaction from step 6, and you should see a warning in the browser console that `withSyncEvent` needs to be used.
